device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py

```diff
@@ -528,49 +536,63 @@ def get_transceiver_change_event(self, timeout=0):
                 buf = self._read_eeprom_devid(x, self.IDENTITY_EEPROM_ADDR, 0x0, 4)
                 if buf is None:
+                    log_info("PORT {0}: unable to get module state".format(x))
                     continue
                 # skip, in case of QSFP28
                 if buf[0] not in "18,19":
                     continue
-                # skip, in case that CMIS < 3.0
+                # skip, in case that CMIS < v3.0
                 if int(buf[1], 16) < 0x30:
                     continue
+                elif int(buf[1], 16) < 0x40:
+                    # As of now, the init sequence does not seem to be necessary for CMIS3
+                    self.mod_inited[x] = True
                 # advance the failure counter if state != ModuleReady
                 if ((int(buf[3], 16) >> 1) & 0x7) != 3:
-                    self.mod_failure[x] += 1
-                # initiate QSFP-DD software reset if failure counter > 2
-                if self.mod_failure[x] > 2:
+                    # fetch the module media type
+                    buf = self._read_eeprom_devid(x, self.IDENTITY_EEPROM_ADDR, 85, 1)
+                    if buf is None:
+                        log_info("PORT {0}: unable to get module media type".format(x))
+                    # perform a software reset only if it's not a DAC (Passive Copper)
+                    elif int(buf[0], 16) != 0x03:
+                        self.mod_failure[x] += 1
+                # initiate QSFP-DD software reset if failure count > 5
+                if self.mod_failure[x] > 5:
                     self.mod_failure[x] = 0
-                    # perform a software reset if it's not a DAC (Passive Copper)
-                    # Note:
                     # This is actually a hotfix for 'Eoptolink EOLD-134HG-02-M6, QSFPDD CMIS v3',
                     # its module state could get stuck at ModuleLowPwr upon reinsertion, and a software
                     # reset is necessary to get it recovered.
-                    buf = self._read_eeprom_devid(x, self.IDENTITY_EEPROM_ADDR, 85, 1)
-                    if (buf is not None) and (int(buf[0], 16) != 0x03):
-                        self._write_byte(x, self.IDENTITY_EEPROM_ADDR, -1, 26, 0x08)
+                    log_info("PORT {0}: issuing a software reset".format(x))
+                    self._write_byte(x, self.IDENTITY_EEPROM_ADDR, -1, 26, 0x08)
 
-                # Monitoring the QSFP-DD initialization state, and reinitiate it if necessary
+                # monitor the QSFP-DD initialization state, and reinitiate it if necessary
+                if self.mod_inited[x]:
+                    continue
                 buf = self._read_eeprom_devid(x, self.IDENTITY_EEPROM_ADDR, self._page_to_flat(202, 0x11), 4)
                 if buf is None:
                     continue
                 err = "".join(buf)
                 if err != '11111111':
```

Comment:
Just for a record, this checker is necessary to prevent redundant CMIS4 init upon warm-reboot, otherwise significant packet drops will be observed.